File: csharp/lesson 4/MergeSort.cs
// https://leetcode.com/problems/sort-an-array/

public class Solution 
{
    public int[] SortArray(int[] nums)
    {
        MergeSort(nums);
        return nums;
    }

    private void MergeSort(int[] nums) 
    {
        Sort(nums, 0, nums.Length - 1, new int[nums.Length]);
    }

    private void Sort(int[] nums, int lo, int hi, int[] temp)
    {
        if (hi <= lo) return;
        int mid = lo + (hi - lo) / 2;
        Sort(nums, lo, mid, temp);
        Sort(nums, mid + 1, hi, temp);
        Merge(nums, lo, mid, hi, temp);
    }

    private void Merge(int[] nums, int lo, int mid, int hi, int[] temp)
    {
        for (int k = lo; k <= hi; k++) 
        {
            temp[k] = nums[k];
        }

        int i = lo, j = mid + 1;
        for (var k = lo; k <= hi; k++)
        {
            if (i > mid) 
            {
                nums[k] = temp[j];
                j++;
            } 
            else if (j > hi) 
            {
                nums[k] = temp[i];
                i++;
            }
            else if (temp[j] < temp[i])
            {
                nums[k] = temp[j];
                j++;
            }
            else
            {
                nums[k] = temp[i];
                i++;
            }
        }
    }
}

File: kotlin/lesson 4/MergeSort.kt
// https://leetcode.com/problems/sort-an-array/

class Solution {
    fun sortArray(nums: IntArray): IntArray {
        mergeSort(nums)
        return nums
    }

    fun mergeSort(nums: IntArray) {
        sort(nums, 0, nums.size - 1, IntArray(nums.size))
    }

    fun sort(nums: IntArray, lo: Int, hi: Int, temp: IntArray) {
        if (hi <= lo) return
        val mid = lo + (hi - lo) / 2
        sort(nums, lo, mid, temp)
        sort(nums, mid + 1, hi, temp)
        merge(nums, lo, mid, hi, temp)
    }

    fun merge(nums: IntArray, lo: Int, mid: Int, hi: Int, temp: IntArray) {
        for (i in lo.. hi) {
            temp[i] = nums[i]
        }

        var i = lo
        var j = mid + 1
        for (k in lo..hi) {
            if (i > mid) {
                nums[k] = temp[j]
                j++
            } else if (j > hi) {
                nums[k] = temp[i]
                i++
            } else if (temp[j] < temp[i]) {
                nums[k] = temp[j]
                j++
            } else {
                nums[k] = temp[i]
                i++
            }
        }
    }
}

